plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -410,6 +408,29 @@ private static <K, V> V getWithValueHolder(Cache<K, AtomicReference<V>> cache, K
         return value;
     }
 
+    private static <K, V> V getIncrementally(
+            Cache<K, AtomicReference<V>> cache,
+            K key,
+            Predicate<V> isSufficient,
+            Supplier<V> loader,
+            Function<V, V> incrementalLoader,
+            BiFunction<V, V, V> merger)
+    {
+        AtomicReference<V> valueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+        V oldValue = valueHolder.get();
+        if (oldValue != null && isSufficient.test(oldValue)) {
+            return oldValue;
+        }
+
+        V newValue = oldValue == null ? loader.get() : incrementalLoader.apply(oldValue);
+
+        if (newValue == null) {
+            throw new InvalidCacheLoadException("Failed to return a value for " + key);
+        }
+
+        return merge(merger, valueHolder, oldValue, newValue);
```

Comment:
i don't like `merge` method because it's semantics is hard to follow

i still think we can inline it without having many more lines

```
      V oldValue = valueHolder.get();
        if (oldValue != null && isSufficient.test(oldValue)) {
            return oldValue;
        }

        V loaded;
        if (oldValue == null) {
            loaded = loader.get();
        }
        else {
            loaded = incrementalLoader.apply(oldValue);
        }
        verifyNotNull(loaded, "loader returned null for %s", key);
        V merged = merger.apply(oldValue, loaded);
        if (!valueHolder.compareAndSet(oldValue, merged)) {
            valueHolder.accumulateAndGet(loaded, merger);
        }

        return merged;
```

Reply:
(note that the above code doesn't throw InvalidCacheLoadException, but i actually don't think this is the exception we should throw, so 🤷 )

Reply:
why do you think we need 
```
V merged = merger.apply(oldValue, loaded);
if (!valueHolder.compareAndSet(oldValue, merged)) {
```
instead of just `valueHolder.accumulateAndGet(loaded, merger);` ?
given the value in the `valueHolder` always grows?

Reply:
the `valueHolder.accumulateAndGet(loaded, merger)` would be sufficient, but would also result in merging being done twice.

Reply:
we could also just do `valueHolder.accumulateAndGet(loaded, merger)` and return the merged result.

that would be correct -- the current value is not worse than oldValue -- but would be harder to reason about. The method correctness would depend on always-growing value, whereas the code I proposed doesn't require this knowledge to be correct.

Reply:
ok, so you don't want to depend on the `always-growing` property. That seems reasonable. Changed to your approach..

Reply:
to be clear: i am aware this is "always growing" and can depend on it if have to, but i favor simplicity, i.e. am happy if i don't need to explain things in the code to the reader.